### ISSUE TITLE
load_and_authorize_resource is not working properly after a before_action find

### ISSUE THREAD

**Issue by Babbz75** [CanCanCommunity/cancancan#703]
### Steps to reproduce

I'm trying to use `load_and_authorize_resource` after using `before_action :find_area, only: %i[show update]` and it's not enforcing authorization. I know my authorizations are setup correctly because I can use `authorize! :update, @area` on each endpoint and it works.

Here is my ability file:

`# frozen_string_literal: true

class Ability
  include CanCan::Ability

  def initialize(user)
    global_ability

    return unless user

    if user.admin?
      admin_ability
    elsif user.user?
      user_ability user
    end
  end

  private

  def global_ability
    can :create, ContactMessage
  end

  def admin_ability
    can :manage, :all
    can :access, :rails_admin # grant access to rails_admin
    can :dashboard, :all # grant access to the dashboard
    cannot :create, User
  end

  def user_ability(user)
    can %i[show create update], Area, business: user.business
  end
end`

Here is my controller:

`# frozen_string_literal: true

module API
  module V1
    class AreasController < ApplicationController
      before_action :find_area, only: %i[show update]
      load_and_authorize_resource

      # GET /api/v1/areas/:uid
      def show
        render_errors('Area not found.', :not_found) if @area.blank?
      end

      # POST /api/v1/areas
      def create
        # authorize! :create, @area
        return render_errors(@area) unless @area.save

        render :show
      end

      # PUT /api/v1/areas/:uid
      def update
        # authorize! :update, @area
        return render_errors(@area) unless @area.update(update_params)

        render :show
      end

      private

      def find_area
        @area = Area.find_by(uid: params[:uid])
      end

      def create_params
        params.require(:area).permit(..params)
      end

      def update_params
        params.require(:area).permit(...params)
      end
    end
  end
end`

### Expected behavior

From this wiki article (https://github.com/CanCanCommunity/cancancan/blob/develop/docs/Authorizing-controller-actions.md) in the override loading section, it says as long as I can the custom load action before the load_and_authorize_resource it should work.

### Actual behavior

My authorization spec fails until I uncomment the the authorization for each individual endpoint.

### System configuration
**Rails version**: rails (~> 6.0.3)

**Ruby version**: ruby 2.6.3p62

**CanCanCan version**: cancancan (3.2.1)

Any help is appreciated!

**Comment by Jyeb** [comment]
Also having an issue with ```load_and_authorize_resource``` not working as expected. When an IDOR attack is performed the authorization seems to be getting bypassed. Authorization checks are still working in functional tests.

Similarly to the above, authorize! checks are working when applied to individual endpoints.

**Comment by ghiculescu** [comment]
@Babbz75 can you share the failing spec?

**Comment by Babbz75** [comment]
@ghiculescu Yes, sorry the spec is fairly basic but it does call a shared example that is for testing authorization.

Here's the spec - 

```
# frozen_string_literal: true

require 'rails_helper'

RSpec.describe 'API::V1::areas#show', type: :request do
  let(:user) { FactoryBot.create(:user) }
  let(:parent_area) { FactoryBot.create(:area, business: user.business) }
  let(:area) { FactoryBot.create(:area, business: user.business, parent_area: parent_area) }

  context 'when user is authenticated' do
    before do
      FactoryBot.create(:area, business: user.business, parent_area: area)
      sign_in user
      get api_v1_area_path(uid: area.uid), params: {}, headers: auth_headers
    end

    context 'and when the area does belong to the users business' do
      it { expect_status :ok }
      it { expect_json_types(area_show_json_types) }
      it { expect_json(area_show_json(area)) }
    end

    context 'and when the area does not belong to the users business' do
      it_behaves_like 'an unauthorized request', :put, '/api/v1/areas/:uid' do
        let(:resource) { FactoryBot.create(:area) }
      end
    end
  end

  context 'when the user is not authenticated' do
    it_behaves_like 'an unauthenticated request', :get, '/api/v1/areas/:uid'
  end
end
```

And the unauthorized request shared example

```
RSpec.shared_examples 'an unauthorized request' do |action, endpoint|
  let(:user) { FactoryBot.create(:user) }

  before do
    sign_in user

    resolved_endpoint = if (defined? resource) && resource.present? && endpoint.include?(':id')
                          endpoint.gsub(':id', resource.id.to_s)
                        else
                          endpoint
                        end

    send(action, resolved_endpoint)
  end

  it { expect_status :forbidden }
  it { expect_json_types(:object) }
  it { expect_json_types(message: :string) }
  it { expect_json(message: 'You are not authorized to access this page.') }
end
```

The shared example basically signs in with a user that is not attached to the resource that was passed in and should be protected.

**Comment by ghiculescu** [comment]
Hmmm. It would be good to simplify that example; there's a lot going on.

I would add `puts(self.class.cancan_resource_class.new(self).send(:resource_instance))` at the start of the `update` action (where your commented out `authorize!` call is), and see what it prints. It should print out the `@area` object. If it doesn't, then something is wrong [here](https://github.com/CanCanCommunity/cancancan/blob/3e7bc1a7baccb199eabc5e6a8b38ce33344525d6/lib/cancan/controller_resource.rb#L40).

**Comment by Babbz75** [comment]
It does for the create but not the show. 
Does this have to do with the special parameter setup in routes using uid instead of id?

**Comment by ghiculescu** [comment]
Okay, a few more debug statements for the `show` action should give you some clues.

- `puts(self.class.cancan_resource_class.new(self).send(:resource_instance))` -> what's this printing currently? nil?
- `puts(self.class.cancan_resource_class.new(self).send(:instance_name))` -> this should print "area", does it?

**Comment by Babbz75** [comment]
Sorry for the delayed responses - this first one currently prints nothing and second prints area
